File: modules/hnsorens/io/gic_v3/main.c
#include "boot_info.h"
#include <modules.h>
#include <api/gic_v3.h>
#include "gic_v3.h"
#include <api/serial_debug.h>
#include <errno.h>
#include <test.h>
#include <type.h>

IMPORT_INTERFACE_ANY(serial, serial);

/* ========================================================================== */
/*                 Module init (optional, not required by vtable)             */
/* ========================================================================== */
int main(boot_info_t *boot_info)
{
	(void)boot_info;

	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;

	int ret = init_global(gicd_base, gicr_base);
	if (ret) {
		serial.printf("init_global failed: %d\n", ret);
		return ret;
	}

	ret = init_core();
	if (ret) {
		serial.printf("init_core failed: %d\n", ret);
		return ret;
	}

	/* Use the EL1 Physical Timer (PPI 30) */
	const irq_vector_t timer_irq = 30;

	/* Configure and enable the timer interrupt */
	ret = configure(timer_irq, IRQ_TRIGGER_LEVEL, 0x80);
	if (ret) {
		serial.printf("configure failed: %d\n", ret);
		return ret;
	}

	ret = set_group(timer_irq, IRQ_GROUP_NON_SECURE);
	if (ret) {
		serial.printf("set_group failed: %d\n", ret);
		return ret;
	}

	ret = enable(timer_irq);
	if (ret) {
		serial.printf("enable failed: %d\n", ret);
		return ret;
	}

	set_core_priority_mask(0xFF);

	/* Program the EL1 physical timer to fire after ~10 ms */
	uint64_t cntfrq;
	__asm__ volatile("mrs %0, cntfrq_el0" : "=r"(cntfrq));
	uint64_t cval;
	__asm__ volatile("mrs %0, cntpct_el0" : "=r"(cval));
	cval += cntfrq / 100; /* ~10 ms */

	__asm__ volatile("msr cntp_cval_el0, %0" : : "r"(cval));

	/* Enable timer (bit 0) */
	__asm__ volatile("msr cntp_ctl_el0, %0" : : "r"(1UL));
	__asm__ volatile("isb");

	/* Poll the acknowledge register until we get the timer interrupt */
	irq_vector_t ack = 1023;
	for (int spin = 0; spin < 10000000; spin++) {
		__asm__ volatile("nop");
		ack = acknowledge();
		if (ack != 1023)
			break;
	}

	if (ack == timer_irq) {
		serial.printf(
			"[GIC Demo] Timer interrupt acknowledged (ID %u). Success!\n",
			ack);
		/* Signal end‑of‑interrupt */
		eoi(ack);
	} else {
		serial.printf(
			"[GIC Demo] acknowledge() returned %u (expected %u). FAILURE.\n",
			ack, timer_irq);
	}

	/* Disable timer */
	__asm__ volatile("msr cntp_ctl_el0, %0" : : "r"(0UL));

	/* Disable the interrupt */
	ret = disable(timer_irq);
	if (ret) {
		serial.printf("disable failed: %d\n", ret);
		return ret;
	}

	return 0;
}

/* ========================================================================== */
/*                 Export the vtable                                          */
/* ========================================================================== */
EXPORT_INTERFACE(interrupt_manager, gic_v3,
		 { .init_global = init_global,
		   .init_core = init_core,
		   .set_core_priority_mask = set_core_priority_mask,
		   .enable = enable,
		   .disable = disable,
		   .configure = configure,
		   .set_group = set_group,
		   .route_to_core = route_to_core,
		   .acknowledge = acknowledge,
		   .end_of_interrupt = eoi,
		   .register_handler = register_handler,
		   .unregister_handler = unregister_handler });

/* ========================================================================== */
/*                              Unit Tests                                    */
/* ========================================================================== */
#ifdef TESTING

TEST(GIC_InitLifecycle)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;

	int ret = init_global(gicd_base, gicr_base);
	EXPECT_EQ(ret, 0);
	ret = init_core();
	EXPECT_EQ(ret, 0);
	TEST_RESULT();
}

TEST(GIC_ConfigureSPI)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	irq_vector_t uart_irq = 33;
	int ret = configure(uart_irq, IRQ_TRIGGER_EDGE, 0x80);
	EXPECT_EQ(ret, 0);

	uint32_t prio_reg =
		*(volatile uint32_t *)(gicd_base + 0x0400 + (uart_irq / 4) * 4);
	uint32_t shift = (uart_irq % 4) * 8;
	EXPECT_EQ((prio_reg >> shift) & 0xFF, 0x80);

	uint32_t cfg_reg = *(volatile uint32_t *)(gicd_base + 0x0C00 +
						  (uart_irq / 16) * 4);
	uint32_t cfg_shift = (uart_irq % 16) * 2;
	EXPECT_EQ((cfg_reg >> cfg_shift) & 0x3, 0x2);
	TEST_RESULT();
}

TEST(GIC_EnablePPI)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	irq_vector_t timer_irq = 27;
	int ret = enable(timer_irq);
	EXPECT_EQ(ret, 0);

	uintptr_t redist_base = gicr_base;
	volatile uint32_t *isenabler0 =
		(volatile uint32_t *)(redist_base + 0x10000 + 0x0100);
	EXPECT_EQ((*isenabler0 >> timer_irq) & 1, 1);
	TEST_RESULT();
}

TEST(GIC_SGISelf)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	/* Enable SGI 0 */
	uintptr_t gicr_sgi = gicr_base + 0x10000;
	*(volatile uint32_t *)(gicr_sgi + 0x0100) = (1U << 0);
	*(volatile uint8_t *)(gicr_sgi + 0x0400) = 0x00;

	/* Unmask CPU priority */
	__asm__ volatile("msr ICC_PMR_EL1, %0" : : "r"(0xFFULL));
	__asm__ volatile("msr ICC_IGRPEN1_EL1, %0" : : "r"(1ULL));
	__asm__ volatile("isb");

	/* Get current MPIDR */
	uint64_t mpidr;
	__asm__ volatile("mrs %0, mpidr_el1" : "=r"(mpidr));
	uint64_t aff0 = mpidr & 0xFF;
	uint64_t aff1 = (mpidr >> 8) & 0xFF;
	uint64_t aff2 = (mpidr >> 16) & 0xFF;
	uint64_t aff3 = (mpidr >> 32) & 0xFF;

	uint64_t sgi_reg = (aff3 << 48) | (aff2 << 32) | (aff1 << 24) |
			   ((uint64_t)(0 & 0x0F) << 24) |
			   ((aff0 & 0xF0) << 16) | (1ULL << (aff0 & 0x0F));
	__asm__ volatile("msr ICC_SGI1R_EL1, %0" : : "r"(sgi_reg));
	__asm__ volatile("dsb sy\n isb");

	irq_vector_t id = acknowledge();
	EXPECT_EQ(id, 0);
	int ret = eoi(id);
	EXPECT_EQ(ret, 0);
	TEST_RESULT();
}

TEST(GIC_RegisterHandler)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	void *dummy_ctx = (void *)0xDEADBEEF;
	isr_handler_t dummy_isr = (isr_handler_t)0x1234;

	int ret = register_handler(27, dummy_isr, dummy_ctx);
	EXPECT_EQ(ret, 0);
	ret = register_handler(27, dummy_isr, dummy_ctx);
	EXPECT_EQ(ret, EBUSY);
	ret = unregister_handler(27);
	EXPECT_EQ(ret, 0);
	TEST_RESULT();
}

TEST(GIC_Bounds)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	int ret = register_handler(2000, (isr_handler_t)0x1234, NULL);
	EXPECT_EQ(ret, EINVAL);
	ret = register_handler(27, NULL, NULL);
	EXPECT_EQ(ret, EINVAL);
	ret = unregister_handler(2000);
	EXPECT_EQ(ret, EINVAL);
	ret = set_group(2000, IRQ_GROUP_NON_SECURE);
	EXPECT_EQ(ret, EINVAL);
	TEST_RESULT();
}

TEST(GIC_MagicCorruption)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	irq_vector_t v = 30;
	isr_handler_t f = (isr_handler_t)0xABCD;
	void *arg = (void *)0x12345678;

	int ret = register_handler(v, f, arg);
	EXPECT_EQ(ret, 0);
	ret = unregister_handler(v);
	EXPECT_EQ(ret, 0);
	TEST_RESULT();
}

TEST(GIC_RouteSPIToCore)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	irq_vector_t spi = 50;
	uint64_t mpidr = get_current_mpidr();
	uint64_t expected = mpidr & ~(1ULL << 31);

	int ret = route_to_core(spi, mpidr);
	EXPECT_EQ(ret, 0);

	uint64_t router_val =
		*(volatile uint64_t *)(gicd_base + 0x6000ULL + (spi * 8));
	EXPECT_EQ(router_val, expected);
	TEST_RESULT();
}

TEST(GIC_RouteSGI_Failure)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	int ret = route_to_core(0, get_current_mpidr());
	EXPECT_EQ(ret, EINVAL);
	TEST_RESULT();
}

TEST(GIC_FullSPILifecycle)
{
	TEST_INIT();
	uintptr_t gicd_base = 0x8000000ULL;
	uintptr_t gicr_base = 0x80A0000ULL;
	init_global(gicd_base, gicr_base);
	init_core();

	irq_vector_t spi = 51;
	uint64_t mpidr = get_current_mpidr();
	isr_handler_t dummy_isr = (isr_handler_t)0x1234;
	void *dummy_ctx = (void *)0xDEADBEEF;

	int ret = route_to_core(spi, mpidr);
	EXPECT_EQ(ret, 0);
	ret = configure(spi, IRQ_TRIGGER_EDGE, 0xA0);
	EXPECT_EQ(ret, 0);
	ret = enable(spi);
	EXPECT_EQ(ret, 0);
	ret = register_handler(spi, dummy_isr, dummy_ctx);
	EXPECT_EQ(ret, 0);

	uint32_t enabler =
		*(volatile uint32_t *)(gicd_base + 0x0100 + (spi / 32) * 4);
	EXPECT_EQ((enabler >> (spi % 32)) & 1, 1);

	uint32_t prio_reg =
		*(volatile uint32_t *)(gicd_base + 0x0400 + (spi / 4) * 4);
	uint32_t prio_shift = (spi % 4) * 8;
	EXPECT_EQ((prio_reg >> prio_shift) & 0xFF, 0xA0);

	uint32_t cfg_reg =
		*(volatile uint32_t *)(gicd_base + 0x0C00 + (spi / 16) * 4);
	uint32_t cfg_shift = (spi % 16) * 2;
	EXPECT_EQ((cfg_reg >> cfg_shift) & 0x3, 0x2);

	ret = unregister_handler(spi);
	EXPECT_EQ(ret, 0);
	ret = disable(spi);
	EXPECT_EQ(ret, 0);
	TEST_RESULT();
}

#endif /* TESTING */

File: modules/hnsorens/io/gic_v3/main.zig
//! GICv3 interrupt controller driver, exporting `InterruptManager`
//! (category "interrupt_manager"). Supports multi-core (GICv3
//! redistributors), SPI routing, and SGI/PPI configuration.
const std = @import("std");
const abi = @import("abi");
const kernel_fmt = @import("kernel_fmt");
const kernel_test = @import("kernel_test");
const mmio = @import("mmio");
const sysreg = @import("sysreg");
const spinlock = @import("spinlock");

pub const serial_if = abi.importInterface(abi.Serial);
pub const exc_if = abi.importInterface(abi.Exceptions);

pub const MAX_INTERRUPT_VECTORS = 1024;
const MAX_CORES_SUPPORTED = 64;

const RegisteredIsr = struct {
    handler: ?abi.IsrHandler = null,
    arg: ?*anyopaque = null,
};

const CoreMap = struct {
    allocated: bool = false,
    mpidr: u64 = 0,
};

var g_gicd: u64 = 0; // Distributor physical base
var g_gicr: u64 = 0; // Redistributor physical base (first frame)
var g_initialized: bool = false;

var s_isr_table: [MAX_INTERRUPT_VECTORS]RegisteredIsr = [_]RegisteredIsr{.{}} ** MAX_INTERRUPT_VECTORS;
var s_core_topology: [MAX_CORES_SUPPORTED]CoreMap = [_]CoreMap{.{}} ** MAX_CORES_SUPPORTED;
var s_isr_lock: spinlock.SpinLock = .{};

// --- Exception vector table ------------------------------------------
//
// Moved out to `hnsorens.arch.exceptions`, which owns VBAR_EL1 and a full
// 16-entry table. This module registers `gicIrqCallback` there for the
// IRQ/FIQ classes (see `main`), so GIC dispatch is just one exception
// callback among the fault/syscall ones -- it no longer needs its own
// vector table or VBAR write.

// --- GICD (Distributor) register map -------------------------------------
//
// Each interrupt-indexed register below is defined once via
// `mmio.IndexedField`, which owns the "N elements packed per register"
// arithmetic (e.g. 4 one-byte priorities per 32-bit IPRIORITYR word) that
// every accessor used to redo by hand. `Isenabler`/`Icenabler` are
// write-1-to-set/write-1-to-clear registers -- a 0 bit there means "leave
// alone", not "clear" -- so they're driven through `writeOneHot` (a raw
// one-hot write) rather than `write` (a read-modify-write), matching their
// hardware semantics.

const GicdCtlr = packed struct(u32) {
    _reserved0: u1 = 0,
    /// GICD_CTLR.EnableGrp1NS (bit 1): enable Group 1 Non-secure interrupts.
    ns_ena_grp1ns: bool = false,
    _reserved1: u2 = 0,
    /// GICD_CTLR.ARE_NS (bit 4): enable affinity-routed (GICv3-style) SPIs.
    ns_are_ns: bool = false,
    _reserved2: u27 = 0,
};

const Gicd = struct {
    const Ctlr = mmio.Reg(GicdCtlr, 0x0000);
    const Igroupr = mmio.IndexedField(bool, 0x0080, u32);
    const Isenabler = mmio.IndexedField(bool, 0x0100, u32);
    const Icenabler = mmio.IndexedField(bool, 0x0180, u32);
    const Ipriorityr = mmio.IndexedField(u8, 0x0400, u32);
    const Icfgr = mmio.IndexedField(u2, 0x0C00, u32);
    const Igrpmodr = mmio.IndexedField(bool, 0x0D00, u32);
    const Irouter = mmio.RegArray(u64, 0x6000, 8);
};

// --- GICR (Redistributor) -- per-core frame offsets ---------------------

const GicrWaker = packed struct(u32) {
    _reserved0: u1 = 0,
    processor_sleep: bool = false,
    children_asleep: bool = false,
    _reserved1: u29 = 0,
};

// SGI/PPI-indexed registers (Isenabler..Igrpmodr) live in the
// redistributor's second 64KB frame -- GICR_SGI_OFFSET is folded into
// their offsets here so callers pass the same per-core base as `Waker`.
const GICR_SGI_OFFSET: u64 = 0x10000;

const Gicr = struct {
    const Waker = mmio.Reg(GicrWaker, 0x0014);
    const Igroupr = mmio.IndexedField(bool, GICR_SGI_OFFSET + 0x0080, u32);
    const Isenabler = mmio.IndexedField(bool, GICR_SGI_OFFSET + 0x0100, u32);
    const Icenabler = mmio.IndexedField(bool, GICR_SGI_OFFSET + 0x0180, u32);
    const Ipriorityr = mmio.IndexedField(u8, GICR_SGI_OFFSET + 0x0400, u32);
    const Icfgr = mmio.IndexedField(u2, GICR_SGI_OFFSET + 0x0C00, u32);
    const Igrpmodr = mmio.IndexedField(bool, GICR_SGI_OFFSET + 0x0D00, u32);
};

// --- CPU interface system registers -------------------------------------

const IccSre = packed struct(u64) {
    /// ICC_SRE_EL1.SRE (bit 0): system register interface enabled.
    sre: bool = false,
    /// ICC_SRE_EL1.DFB (bit 1): disable IRQ/FIQ bypass.
    dfb: bool = false,
    /// ICC_SRE_EL1.DIB (bit 2): disable IRQ/FIQ bypass (the other one).
    dib: bool = false,
    _reserved: u61 = 0,
};

const IccIgrpen1 = packed struct(u64) {
    /// ICC_IGRPEN1_EL1.Enable (bit 0): enable Group 1 interrupts.
    enable: bool = false,
    _reserved: u63 = 0,
};

const Icc = struct {
    const Sre = sysreg.Reg(IccSre, "ICC_SRE_EL1");
    const Pmr = sysreg.Reg(u64, "ICC_PMR_EL1");
    const Bpr1 = sysreg.Reg(u64, "ICC_BPR1_EL1");
    const Ctlr = sysreg.Reg(u64, "ICC_CTLR_EL1");
    const Igrpen1 = sysreg.Reg(IccIgrpen1, "ICC_IGRPEN1_EL1");
    const Iar1 = sysreg.Reg(u64, "ICC_IAR1_EL1");
    const Eoir1 = sysreg.Reg(u64, "ICC_EOIR1_EL1");
};

const Mpidr = sysreg.Reg(u64, "mpidr_el1");

fn mpidrToCoreId(mpidr: u64) u32 {
    const aff0: u32 = @truncate(mpidr & 0xFF);
    const aff1: u32 = @truncate((mpidr >> 8) & 0xFF);
    return aff0 + aff1; // single cluster => linear index
}

pub fn readMpidr() u64 {
    return Mpidr.read();
}

fn getCurrentGicrBase() u64 {
    const cid = mpidrToCoreId(readMpidr());
    return g_gicr + @as(u64, cid) * 0x20000;
}

fn gicv3InitRedistributor() c_int {
    const base = getCurrentGicrBase();
    Gicr.Waker.modify(base, .{ .processor_sleep = false });
    asm volatile ("dsb sy");

    var i: u32 = 0;
    while (i < 1000) : (i += 1) {
        if (!Gicr.Waker.read(base).children_asleep) return 0;
    }
    return abi.EIO;
}

fn gicv3InitCpuInterfaceNs() c_int {
    Icc.Sre.modify(.{ .sre = true, .dfb = true, .dib = true });
    asm volatile ("isb");

    Icc.Pmr.write(0xFF);
    Icc.Bpr1.write(0);
    Icc.Ctlr.write(0);
    Icc.Igrpen1.write(.{ .enable = true });
    asm volatile ("dsb sy");
    asm volatile ("isb");
    return 0;
}

fn iccWritePmrEl1(val: u64) void {
    Icc.Pmr.write(val);
    asm volatile ("isb");
}

fn iccReadIar1El1() u64 {
    return Icc.Iar1.read();
}

fn iccWriteEoir1El1(val: u64) void {
    Icc.Eoir1.write(val);
    asm volatile ("isb");
}

// --- 1. System-wide initialization --------------------------------------

pub fn initGlobal(d_base: u64, r_base: u64) callconv(.c) c_int {
    if (d_base == 0 or r_base == 0) return abi.EINVAL;
    if (g_initialized) return abi.EBUSY;

    g_gicd = d_base;
    g_gicr = r_base;

    Gicd.Ctlr.modify(g_gicd, .{ .ns_are_ns = true, .ns_ena_grp1ns = true });
    asm volatile ("dsb sy");

    g_initialized = true;
    return 0;
}

// --- 2. Per-core initialization ------------------------------------------

pub fn initCore() callconv(.c) c_int {
    if (!g_initialized) return abi.EIO;

    const mpidr = readMpidr();
    const core_id = mpidrToCoreId(mpidr);
    if (core_id >= MAX_CORES_SUPPORTED) return abi.EINVAL;

    s_core_topology[core_id].mpidr = mpidr;
    s_core_topology[core_id].allocated = true;

    var ret = gicv3InitRedistributor();
    if (ret != 0) return ret;
    ret = gicv3InitCpuInterfaceNs();
    if (ret != 0) return ret;

    return 0;
}

// --- 3. Priority mask (per-core) -----------------------------------------

pub fn setCorePriorityMask(mask: u32) callconv(.c) c_int {
    iccWritePmrEl1(mask);
    return 0;
}

// --- 4. Interrupt enable / disable ---------------------------------------

pub fn enable(vector: u32) callconv(.c) c_int {
    if (vector >= MAX_INTERRUPT_VECTORS) return abi.EINVAL;

    if (vector < 32) {
        Gicr.Isenabler.writeOneHot(getCurrentGicrBase(), vector);
    } else {
        Gicd.Isenabler.writeOneHot(g_gicd, vector);
    }
    asm volatile ("dsb sy");
    return 0;
}

pub fn disable(vector: u32) callconv(.c) c_int {
    if (vector >= MAX_INTERRUPT_VECTORS) return abi.EINVAL;

    if (vector < 32) {
        Gicr.Icenabler.writeOneHot(getCurrentGicrBase(), vector);
    } else {
        Gicd.Icenabler.writeOneHot(g_gicd, vector);
    }
    asm volatile ("dsb sy");
    return 0;
}

// --- 5. Interrupt configuration (trigger, priority) ----------------------

pub fn configure(vector: u32, trigger: abi.IrqTrigger, priority: u32) callconv(.c) c_int {
    if (vector >= MAX_INTERRUPT_VECTORS) return abi.EINVAL;

    const trigger_val: u2 = if (trigger == .edge) 0b10 else 0b00;
    const prio: u8 = @truncate(priority);

    if (vector < 32) {
        const base = getCurrentGicrBase();
        Gicr.Ipriorityr.write(base, vector, prio);
        Gicr.Icfgr.write(base, vector, trigger_val);
        Gicr.Igroupr.write(base, vector, true);
        Gicr.Igrpmodr.write(base, vector, false); // Non-Secure.
    } else {
        Gicd.Ipriorityr.write(g_gicd, vector, prio);
        Gicd.Icfgr.write(g_gicd, vector, trigger_val);
        Gicd.Igroupr.write(g_gicd, vector, true);
        Gicd.Igrpmodr.write(g_gicd, vector, false); // Non-Secure.
    }

    asm volatile ("dsb sy" ::: .{ .memory = true });
    return 0;
}

// --- 6. Group assignment --------------------------------------------------

pub fn setGroup(vector: u32, group: abi.IrqGroup) callconv(.c) c_int {
    if (vector >= MAX_INTERRUPT_VECTORS) return abi.EINVAL;

    const non_secure = group == .non_secure;
    if (vector < 32) {
        Gicr.Igroupr.write(getCurrentGicrBase(), vector, non_secure);
    } else {
        Gicd.Igroupr.write(g_gicd, vector, non_secure);
    }
    asm volatile ("dsb sy" ::: .{ .memory = true });
    return 0;
}

// --- 7. SPI routing to a core ---------------------------------------------

pub fn routeToCore(vector: u32, mpidr: u64) callconv(.c) c_int {
    if (vector < 32 or vector >= MAX_INTERRUPT_VECTORS) return abi.EINVAL;

    var found = false;
    for (&s_core_topology) |*core| {
        if (core.allocated and core.mpidr == mpidr) {
            found = true;
            break;
        }
    }
    if (!found) return abi.EINVAL;

    // Clear IRM bit (bit 31) to force unicast delivery.
    const routing_val = mpidr & ~(@as(u64, 1) << 31);
    Gicd.Irouter.write(g_gicd, vector, routing_val);
    asm volatile ("dsb sy" ::: .{ .memory = true });
    return 0;
}

// --- 8. Acknowledge & End Of Interrupt ------------------------------------

pub fn acknowledge() callconv(.c) u32 {
    return @truncate(iccReadIar1El1() & 0xFFFFFFFF);
}

pub fn eoi(vector: u32) callconv(.c) c_int {
    iccWriteEoir1El1(vector);
    return 0;
}

// --- 9. Handler registration -----------------------------------------------

pub fn registerHandler(vector: u32, handler: ?abi.IsrHandler, arg: ?*anyopaque) callconv(.c) c_int {
    if (vector >= MAX_INTERRUPT_VECTORS or handler == null) return abi.EINVAL;

    s_isr_lock.lock();
    if (s_isr_table[vector].handler != null) {
        s_isr_lock.unlock();
        return abi.EBUSY;
    }
    s_isr_table[vector].handler = handler;
    s_isr_table[vector].arg = arg;
    s_isr_lock.unlock();

    const ret = enable(vector);
    if (ret != 0) {
        s_isr_lock.lock();
        s_isr_table[vector].handler = null;
        s_isr_table[vector].arg = null;
        s_isr_lock.unlock();
    }
    return ret;
}

pub fn unregisterHandler(vector: u32) callconv(.c) c_int {
    if (vector >= MAX_INTERRUPT_VECTORS) return abi.EINVAL;

    const ret = disable(vector);
    if (ret != 0) return ret;

    s_isr_lock.lock();
    s_isr_table[vector].handler = null;
    s_isr_table[vector].arg = null;
    s_isr_lock.unlock();
    return 0;
}

// --- 10. IRQ dispatch (registered with the exceptions module) -------------

/// Registered for `.irq` and `.fiq` with `hnsorens.arch.exceptions`. The
/// trap frame is unused -- GIC state comes from the CPU interface system
/// registers (IAR1/EOIR1), not the saved GPRs.
fn gicIrqCallback(frame: *abi.TrapFrame, origin: abi.ExceptionOrigin, arg: ?*anyopaque) callconv(.c) abi.ExceptionOutcome {
    _ = frame;
    _ = origin;
    _ = arg;

    const iar = iccReadIar1El1();
    const id: u32 = @truncate(iar & 0xFFFFFFFF);

    if (id == 1023) return .handled; // spurious

    if (id < MAX_INTERRUPT_VECTORS) {
        s_isr_lock.lock();
        const handler = s_isr_table[id].handler;
        const arg_ctx = s_isr_table[id].arg;
        s_isr_lock.unlock();

        if (handler) |h| {
            h(arg_ctx);
        } else {
            kernel_fmt.print(serial_if, "Unhandled Interrupt ID: {d}\n", .{id});
        }
    } else {
        kernel_fmt.print(serial_if, "Interrupt ID out of bounds: {d}\n", .{id});
    }

    iccWriteEoir1El1(iar);
    return .handled;
}

var s_irq_hook_installed: bool = false;

/// Registers `gicIrqCallback` for the IRQ and FIQ exception classes with
/// the exceptions module. Runs once, before any GIC test or the timer
/// module's init.
pub fn main(boot_info_ptr: *anyopaque) void {
    _ = boot_info_ptr;
    if (s_irq_hook_installed) return;
    _ = exc_if.register_handler(.irq, &gicIrqCallback, null);
    _ = exc_if.register_handler(.fiq, &gicIrqCallback, null);
    s_irq_hook_installed = true;
}

// --- 11. Helper to read current MPIDR (useful for the OS) ------------------

pub fn getCurrentMpidr() u64 {
    return readMpidr();
}

comptime {
    abi.exportInterface("gic_v3", abi.InterruptManager, .{
        .init_global = initGlobal,
        .init_core = initCore,
        .set_core_priority_mask = setCorePriorityMask,
        .enable = enable,
        .disable = disable,
        .configure = configure,
        .set_group = setGroup,
        .route_to_core = routeToCore,
        .acknowledge = acknowledge,
        .end_of_interrupt = eoi,
        .register_handler = registerHandler,
        .unregister_handler = unregisterHandler,
    });
}

comptime {
    _ = @import("test.zig");
}
